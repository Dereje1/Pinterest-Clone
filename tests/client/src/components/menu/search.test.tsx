/**
 * @jest-environment jsdom
 */
import React from 'react';
import { EnzymePropSelector, shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import * as redux from 'react-redux';
import Search from '../../../../../client/src/components/menu/search';

// Mock redux hooks
jest.mock('react-redux', () => ({
  ...jest.requireActual('react-redux'), // use actual for all non-hook parts
  useSelector: jest.fn((() => ({ term: null, tagSearch: false }))),
}));

jest.useFakeTimers();
describe('The search component', () => {
  let props: React.ComponentProps<typeof Search>;
  beforeEach(() => {
    props = {
      searchUpdate: jest.fn(),
      pathname: '/',
      isShowing: false,
      openSearch: jest.fn(),
      closeSearch: jest.fn(),
    };
    global.scrollTo = jest.fn();
  });

  test('Will not render anything if at /pins path', () => {
    const updatedProps = { ...props, pathname: '/pins' };
    const wrapper = shallow(<Search {...updatedProps} />);
    expect(wrapper.isEmptyRender()).toBe(true);
    expect(props.closeSearch).toHaveBeenCalled();
  });

  test('Will render search bar if selected by the user', () => {
    const updatedProps = { ...props, isShowing: true };
    const wrapper = shallow(<Search {...updatedProps} />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  test('will update the search in state and the redux store on change', () => {
    const updatedProps = { ...props, isShowing: true };
    const wrapper = shallow(<Search {...updatedProps} />);
    let searchInput: EnzymePropSelector = wrapper.find('ForwardRef(InputBase)');
    searchInput.props().onChange({ target: { value: 'abc' } });
    jest.advanceTimersByTime(1000);
    expect(props.searchUpdate).toHaveBeenCalledWith('abc');
    searchInput = wrapper.find('ForwardRef(InputBase)');
    expect(searchInput.props().value).toBe('abc');
  });
  test('will clear the search in state and the redux store on click', () => {
    const updatedProps = { ...props, isShowing: true };
    const wrapper = shallow(<Search {...updatedProps} />);
    let searchInput: EnzymePropSelector = wrapper.find('ForwardRef(InputBase)');
    // first call to change search terms
    searchInput.props().onChange({ target: { value: 'abc' } });
    jest.advanceTimersByTime(1000);
    expect(props.searchUpdate).toHaveBeenNthCalledWith(1, 'abc');
    searchInput = wrapper.find('ForwardRef(InputBase)');
    expect(searchInput.props().value).toBe('abc');
    // second call to clear search terms
    const clearButton = wrapper.find({ id: 'clear-search' });
    clearButton.props().onClick();
    searchInput = wrapper.find('ForwardRef(InputBase)');
    expect(searchInput.props().value).toBe('');
    expect(props.searchUpdate).toHaveBeenNthCalledWith(2, '');
  });

  test('will clear the search in state and the redux store if arrow back is clicked', () => {
    const updatedProps = { ...props, isShowing: true };
    const wrapper = shallow(<Search {...updatedProps} />);
    let searchInput: EnzymePropSelector = wrapper.find('ForwardRef(InputBase)');
    // first call to change search terms
    searchInput.props().onChange({ target: { value: 'abc' } });
    jest.advanceTimersByTime(1000);
    expect(props.searchUpdate).toHaveBeenNthCalledWith(1, 'abc');
    searchInput = wrapper.find('ForwardRef(InputBase)');
    expect(searchInput.props().value).toBe('abc');
    // second call to clear search terms
    const clearButton = wrapper.find({ id: 'back' });
    clearButton.props().onClick();
    searchInput = wrapper.find('ForwardRef(InputBase)');
    expect(searchInput.props().value).toBe('');
    expect(props.searchUpdate).toHaveBeenNthCalledWith(2, '');
  });

  test('Will keep search open with prev value if user comes back to home', () => {
    // start with root path and update search val
    const updatedProps = { ...props, isShowing: true };
    const wrapper = shallow(<Search {...updatedProps} />);
    // update search val
    let searchInput: EnzymePropSelector = wrapper.find('ForwardRef(InputBase)');
    searchInput.props().onChange({ target: { value: 'abc' } });
    expect(wrapper.isEmptyRender()).toBe(false);
    // go back to a non searchable path
    wrapper.setProps({ ...props, pathname: '/pins' });
    expect(wrapper.isEmptyRender()).toBe(true);
    expect(props.closeSearch).toHaveBeenCalled();
    // then go back to root path again
    wrapper.setProps({ ...props, pathname: '/', isShowing: true });
    searchInput = wrapper.find('ForwardRef(InputBase)');
    expect(props.openSearch).toHaveBeenCalled();
    expect(searchInput.props().value).toBe('abc');
  });

  test('will use the tag terms to search if change in redux store', () => {
    jest
      .spyOn(redux, 'useSelector')
      .mockImplementationOnce(() => ({ term: 'tag search', tagSearch: true }));
    const wrapper = shallow(<Search {...props} isShowing />);
    jest.advanceTimersByTime(1000);
    const searchInput = wrapper.find('ForwardRef(InputBase)');
    expect(props.searchUpdate).toHaveBeenCalledWith('tag search');
    expect(props.openSearch).toHaveBeenCalled();
    expect(searchInput.props().value).toBe('tag search');
    expect(global.scrollTo).toHaveBeenCalled();
  });
});
