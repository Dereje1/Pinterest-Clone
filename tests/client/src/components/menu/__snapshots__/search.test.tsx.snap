// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The search component Will render only search Icon if not selected by the user 1`] = `
<div
  className="search"
>
  <ForwardRef(Tooltip)
    placement="bottom"
    title="Search by description, owner or tags"
  >
    <Memo(ForwardRef(SearchIcon))
      fontSize="large"
      onClick={[MockFunction]}
      sx={
        {
          "cursor": "pointer",
          "justifySelf": "flex-end",
        }
      }
    />
  </ForwardRef(Tooltip)>
</div>
`;

exports[`The search component Will render search bar if selected by the user 1`] = `
<div
  style={
    {
      "alignItems": "center",
      "display": "flex",
      "justifyContent": "space-around",
      "marginLeft": "auto",
      "marginRight": "auto",
    }
  }
>
  <ForwardRef(Paper)
    component="form"
    elevation={3}
    sx={
      {
        "alignItems": "center",
        "display": "flex",
        "marginBottom": "auto",
        "marginTop": "auto",
        "p": "2px 4px",
        "width": 360,
      }
    }
    variant="elevation"
  >
    <ForwardRef(IconButton)
      aria-label="menu"
      id="back"
      onClick={[Function]}
      sx={
        {
          "p": "10px",
        }
      }
    >
      <Memo(ForwardRef(ArrowBackIcon)) />
    </ForwardRef(IconButton)>
    <ForwardRef(InputBase)
      autoFocus={true}
      inputProps={
        {
          "aria-label": "search",
        }
      }
      onChange={[Function]}
      placeholder="Search..."
      sx={
        {
          "height": "100%",
          "marginLeft": 1,
          "width": "100%",
        }
      }
      value=""
    />
    <ForwardRef(IconButton)
      id="clear-search"
      onClick={[Function]}
      sx={
        {
          "cursor": "pointer",
          "p": "10px",
          "visibility": "hidden",
        }
      }
      type="button"
    >
      <Memo(ForwardRef(HighlightOffIcon)) />
    </ForwardRef(IconButton)>
  </ForwardRef(Paper)>
</div>
`;
